test/common/secret/sds_api_test.cc

```diff
@@ -115,6 +115,92 @@ TEST_F(SdsApiTest, DynamicTlsCertificateUpdateSuccess) {
+TEST_F(SdsApiTest, Delta) {
+  Protobuf::RepeatedPtrField<envoy::api::v2::Resource> resources;
+  envoy::api::v2::auth::Secret secret;
+  secret.set_name("secret_1");
+  auto* resource = resources.Add();
+  resource->mutable_resource()->PackFrom(secret);
+  resource->set_name("secret_1");
+  resource->set_version("version1");
+
+  NiceMock<Server::MockInstance> server;
+  NiceMock<Init::MockManager> init_manager;
+  envoy::api::v2::core::ConfigSource config_source;
+  PartialMockSds sds(server, *api_, init_manager, config_source);
+  EXPECT_CALL(sds, onConfigUpdate(_, "version1"));
+  sds.SdsApi::onConfigUpdate(resources, {}, "ignored");
+
+  // An attempt to remove a resource logs an error, but otherwise just carries on (ignoring the
+  // removal attempt).
+  resource->set_version("version2");
+  EXPECT_CALL(sds, onConfigUpdate(_, "version2"));
+  Protobuf::RepeatedPtrField<std::string> removals;
+  *removals.Add() = "route_0";
+  sds.SdsApi::onConfigUpdate(resources, removals, "ignored");
+}
+
+// Tests SDS's use of the delta variant of onConfigUpdate().
+TEST_F(SdsApiTest, DeltaUpdateSuccess) {
+  NiceMock<Server::MockInstance> server;
+  NiceMock<Init::MockManager> init_manager;
+  envoy::api::v2::core::ConfigSource config_source;
+  TlsCertificateSdsApi sds_api(
+      server.localInfo(), server.dispatcher(), server.random(), server.stats(),
+      server.clusterManager(), init_manager, config_source, "abc.com", []() {}, *api_);
+
+  NiceMock<Secret::MockSecretCallbacks> secret_callback;
+  auto handle =
+      sds_api.addUpdateCallback([&secret_callback]() { secret_callback.onAddOrUpdateSecret(); });
+
+  std::string yaml =
+      R"EOF(
+  name: "abc.com"
+  tls_certificate:
+    certificate_chain:
+      filename: "{{ test_rundir }}/test/extensions/transport_sockets/tls/test_data/selfsigned_cert.pem"
+    private_key:
+      filename: "{{ test_rundir }}/test/extensions/transport_sockets/tls/test_data/selfsigned_key.pem"
+    )EOF";
+  envoy::api::v2::auth::Secret typed_secret;
+  MessageUtil::loadFromYaml(TestEnvironment::substitute(yaml), typed_secret);
+  Protobuf::RepeatedPtrField<envoy::api::v2::Resource> secret_resources;
+  secret_resources.Add()->mutable_resource()->PackFrom(typed_secret);
+
+  EXPECT_CALL(secret_callback, onAddOrUpdateSecret());
+  sds_api.onConfigUpdate(secret_resources, {}, "");
+
+  Ssl::TlsCertificateConfigImpl tls_config(*sds_api.secret(), *api_);
+  const std::string cert_pem =
+      "{{ test_rundir }}/test/extensions/transport_sockets/tls/test_data/selfsigned_cert.pem";
+  EXPECT_EQ(TestEnvironment::readFileToStringForTest(TestEnvironment::substitute(cert_pem)),
+            tls_config.certificateChain());
+
+  const std::string key_pem =
+      "{{ test_rundir }}/test/extensions/transport_sockets/tls/test_data/selfsigned_key.pem";
+  EXPECT_EQ(TestEnvironment::readFileToStringForTest(TestEnvironment::substitute(key_pem)),
+            tls_config.privateKey());
+
+  handle->remove();
+}
+
```

Comment:
Should we also have something similar for RDS? I think you can basically verify the `onConfigUpdate` chaining without relying on any internal implementation details of RDS.

Reply:
I actually tried RDS before doing SDS. The amount of machinery needed to instantiate an object was going to be ridiculous, especially because I think it was going to need to go through the new ConfigProvider setup; I'm not even sure how that would work. I think it would be cleaner to just use the mostly-copied test I originally had.

Reply:
Hmm, this seems a code smell if we can't instantiate a module like the RDS subscription without a lot of setup. Is there no easy way to mock out all this ConfigProvider stuff? I feel we have some structural issues if we can't instantiate some module like RDS subscription and test individual functions cheaply.

Reply:
CC @AndresGuedez for possible insight into how RDS can made more testable.

Reply:
@fredlas just a minor clarification (or maybe I'm misunderstanding your statement) -- RDS currently uses its own `*ConfigProvider*` set of classes which are the inspiration for the new `ConfigProvider` framework; RDS hasn't yet been ported to this generalized framework on which SRDS is built.

With that said, the testing approach for both the RDS `ConfigProvider` and the generalized framework could be similar. My suggestion would be to have a test peer to [`RouteConfigProviderManagerImpl`](https://github.com/envoyproxy/envoy/blob/master/source/common/router/rds_impl.h#L185) which can inject a factory to use for creating an `RdsRouteConfigSubscription`. When `createRdsRouteConfigProvider()` is called, a `RouteConfigProvider` associated with the injected `MockRdsRouteConfigSubscription` would then be returned.

Reply:
Thanks for the summary, Andres. That is exactly the complexity that seems too much for just testing a passthrough function!

Reply:
We discussed this at our incremental xDS sync meeting today. Since we have e2e integration test coverage of this functionality, I think we can skip the unit tests since they appear to be too PITA.